packages/contentstack-import/src/import/modules/base-class.ts

```diff
@@ -0,0 +1,218 @@
+  includeParamOnCompletion?: boolean;
+  serializeData?: (input: ApiOptions) => any;
+};
+
+export type EnvType = {
+  processName: string;
+  totalCount?: number;
+  apiContent: Record<string, any>[];
+  concurrencyLimit?: number;
+  apiParams?: ApiOptions;
+};
+
+export type CustomPromiseHandlerInput = {
+  index: number;
+  batchIndex: number;
+  apiParams?: ApiOptions;
+  isLastRequest: boolean;
+};
+
+export type CustomPromiseHandler = (input: CustomPromiseHandlerInput) => Promise<any>;
+
+export default abstract class BaseClass {
+  readonly client: Stack;
+
+  public importConfig: typeof configType & AdditionalKeys;
+
+  public modulesConfig: typeof configType.modules;
+
+  constructor({ importConfig, stackAPIClient }) {
+    this.client = stackAPIClient;
+    this.importConfig = importConfig;
+    this.modulesConfig = importConfig.modules;
+  }
+
+  get stack(): Stack {
+    return this.client;
+  }
+
+  /**
+   * @method delay
+   * @param {number} ms number
+   * @returns {Promise} Promise<void>
+   */
+  delay(ms: number): Promise<void> {
+    /* eslint-disable no-promise-executor-return */
+    return new Promise((resolve) => setTimeout(resolve, ms <= 0 ? 0 : ms));
+  }
+
+  /**
+   * @method makeConcurrentCall
+   * @param {Record<string, any>} env EnvType
+   * @param {CustomPromiseHandler} promisifyHandler CustomPromiseHandler
+   * @param {boolean} logBatchCompletionMsg boolean
+   * @returns {Promise} Promise<void>
+   */
+  makeConcurrentCall(
+    env: EnvType,
+    promisifyHandler?: CustomPromiseHandler,
+    logBatchCompletionMsg = true,
+  ): Promise<void> {
+    const { processName, apiContent, apiParams, concurrencyLimit = this.importConfig.modules.apiConcurrency } = env;
+
+    /* eslint-disable no-async-promise-executor */
+    return new Promise(async (resolve) => {
+      let batchNo = 0;
+      let isLastRequest = false;
+      const batches: Array<Record<string, any>> = chunk(apiContent, concurrencyLimit);
+
+      /* eslint-disable no-promise-executor-return */
+      if (isEmpty(batches)) return resolve();
+
+      for (const [batchIndex, batch] of entries(batches)) {
+        batchNo += 1;
+        const allPromise = [];
+        const start = Date.now();
+
+        for (const [index, element] of entries(batch)) {
+          let promise = Promise.resolve();
+          isLastRequest = isEqual(last(batch as ArrayLike<any>), element) && isEqual(last(batches), batch);
+
+          if (promisifyHandler instanceof Function) {
+            promise = promisifyHandler({
+              apiParams,
+              isLastRequest,
+              index: Number(index),
+              batchIndex: Number(batchIndex),
+            });
+          } else if (apiParams) {
+            apiParams.apiData = element;
+            promise = this.makeAPICall(apiParams, isLastRequest);
+          }
+
+          allPromise.push(promise);
+        }
+
+        /* eslint-disable no-await-in-loop */
+        await Promise.allSettled(allPromise);
```

Comment:
@antonyagustine Can we have a provision to handle the errors caught from this Promise call?

Reply:
@netrajpatel We capture all the errors in each request. Here you can find the link for error handling.
https://github.com/contentstack/cli/pull/597/files#diff-0c1e791feb0ad5ec979aadf45f8669196546cc91d3140ec6c7fd5d70d01be9e2R180